File: harlie_vision/src/image_to_point_cloud.py
#!/usr/bin/env python

import roslib
roslib.load_manifest("harlie_vision")
import rospy
import struct

from sensor_msgs.msg import Image
from sensor_msgs.msg import PointCloud
from geometry_msgs.msg import Point32

def image_callback(img, pub):
	cloud = PointCloud()
	cloud.header.stamp = rospy.Time.now()
	cloud.header.frame_id = "/vision_cloud"
	point = Point32(300*0.05, -300*0.05, 0.1)
	for i in range(0, img.height):
		point.y = -300*0.05
		for j in range(0, img.width):
			data = img.data[img.step*i+j]
			val = struct.unpack("B", data)[0]
			if(val/255.0 > 0.1):
				cloud.points.append(Point32(point.x, point.y, point.z))
			point.y += 0.05
		point.x -= 0.05
	pub.publish(cloud)	

if __name__ == "__main__":
	rospy.init_node("image_to_point_cloud")
	pub = rospy.Publisher("vision_cloud", PointCloud) 
	rospy.Subscriber("plan_view", Image, image_callback, pub)
	rospy.spin()

File: cwru_experimental/cwru_vision/src/image_to_point_cloud.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <sensor_msgs/PointCloud.h>
#include <image_transport/image_transport.h>
#include <geometry_msgs/Point32.h>

ros::Publisher pub_;
sensor_msgs::PointCloud cloud;

void image_callback(const sensor_msgs::ImageConstPtr& img);

int main(int argc, char *argv[]) {
    ros::init(argc, argv, "image_to_point_cloud");
    ros::NodeHandle nh_;
    pub_ = nh_.advertise<sensor_msgs::PointCloud>("vision_cloud",1); 
    image_transport::ImageTransport it_(nh_);
    image_transport::Subscriber image_subscriber = it_.subscribe("plan_view", 1, image_callback);
    ros::spin();
}

void image_callback(const sensor_msgs::ImageConstPtr& img) {
    cloud.header.stamp = ros::Time::now();
    cloud.header.frame_id = "vision_cloud";
    geometry_msgs::Point32 temp = geometry_msgs::Point32(300*0.05, -300*0.05, 0);
    for (int i = 0; i < img->height; i++) {
	for (int j = 0; j < img->width; j++) {
	    if(img->data[image->step*i + j] > 0) {
		cloud.points << temp;
	    }
	    temp.y += 0.05
	}
	temp.x -= 0.05;
    }
}
